src/Symfony/Component/Security/Http/EventListener/IsGrantedAttributeListener.php

```diff
@@ -68,7 +77,7 @@ public function onKernelControllerArguments(ControllerArgumentsEvent $event)
                 }
 
                 $accessDeniedException = new AccessDeniedException($message);
-                $accessDeniedException->setAttributes($attribute->attributes);
+                $accessDeniedException->setAttributes($attribute->attributes instanceof Expression ? (string) $attribute->attributes : $attribute->attributes);
```

Comment:
I think this might be a bug in the `AccessDeniedException` class. There are several ways `setAttributes` could receive an instance of `Expression`, eg
```php
$this->denyAccessUnlessGranted(new Expression('"ROLE_ADMIN" in role_names'))
```
On Symfony 5.4 or lower there are no type hints so the object gets cast to an array:

https://github.com/symfony/symfony/blob/ee2eb1c2ba82483da7ea566b7ce830b4eb1200f9/src/Symfony/Component/Security/Core/Exception/AccessDeniedException.php#L37-L43
```
array(1) {
  ["*expression"]=>
  string(15) "some expression"
}
```

On 6.0 or higher, because of the added type hints, it gets cast into a string instead.
https://github.com/symfony/symfony/blob/c090b9aeefc3c539b3984c211a4b465d5e257b2c/src/Symfony/Component/Security/Core/Exception/AccessDeniedException.php#L34-L37

I think the `setAttributes` method should accept an `Expression` as well and do the casting to a string. I can open a PR if needed.